### ISSUE TITLE
setDoc with option merge: true still overwrites old document

### ISSUE THREAD

**Issue by raaaahman** [firebase/firebase-js-sdk#7033]
### Environment

  * Operating System version: Ubuntu 22.04.1
  * Browser version: Firefox 109.0.1
  * Firebase SDK version: 9.15.0
  * Firebase Product: Firestore

### Description

I am trying to use the `setDoc` function from Firebase Wab SDK v9 with the `{ merge: true }` option like stated in the documentation: https://firebase.google.com/docs/firestore/manage-data/add-data#set_a_document

However, the old document is overwritten regardless of the value of this option.

#### Steps to reproduce:

1. create a new Firebase project
2. activate Firestore for this project
3. create a new TypeScript project (use `npm create vite@latest`for example)
4. configure your Firebase app
5. Create a `VariableStorage` class in your project, and instantiate it passing a `Firestore` and `Auth` instances to its constructor.
6. See code below

#### Relevant Code:

I have a unit test:

```typescript
        it('should not delete other variables', async () => {
            const freddy = testEnv.authenticatedContext(TEST_UID)
            // @ts-ignore
            freddy.auth = () => ({
                currentUser: {
                    uid: TEST_UID
                }
            })
            // @ts-ignore
            const variableStorage = new VariableStorage(freddy)

            variableStorage.set('$somebody_else', 'Elliot')

            expect(await variableStorage.get('$has_met_georges')).toEqual(TEST_VARIABLES.$has_met_georges)
        })
```

The code below make it fail:

```typescript
    public set(key: string, value: VariableType, seasonNb = 1) {
        if(!this.auth.currentUser) {
            throw new Error("Cannot save variable for an unauthenticated user")
        }

        this.variables[key] = value
        try {
            setDoc(
                doc(this.store, VariableStorage.getDocumentPath(this.auth.currentUser.uid, seasonNb)),
                {
                    [key]: value
                },
                {
                    merge: false
                }
            )
        } catch(e) {

        }
    }
```

However, the code below make it pass:

```typescript
    public set(key: string, value: VariableType, seasonNb = 1) {
        if(!this.auth.currentUser) {
            throw new Error("Cannot save variable for an unauthenticated user")
        }

        this.variables[key] = value
        try {
            updateDoc(
                doc(this.store, VariableStorage.getDocumentPath(this.auth.currentUser.uid, seasonNb)),
                {
                    [key]: value
                }
            )
        } catch(e) {

        }
    }
```

The issue is, if I have to use the `updateDoc`, function to validate my other requirements, I would have to also start by fetching a snapshot with `getDoc`, verify it exists, and use the `addDoc` method if it doesn't.

#### Related issues

See this [Stack Overflow question 75160998](https://stackoverflow.com/questions/75160998/firestore-setdoc-function-not-merging-new-items-on-react-js/75439887)

**Comment by milaGGL** [comment]
Hi @raaaahman, thank your for reporting this issue. 

Based on the code provided, I assume you are trying to set the "somebody_else" to "Elliot" with merge, have tried {merge:true} and {merge:false}, but "has_met_georges" got effected on both cases, which is not the expected behaviour: 
 https://github.com/firebase/firebase-js-sdk/blob/dd9acd053e9372e28eafe48cf1c5d567b03b8b9f/packages/firestore/test/integration/api/database.test.ts#L178 

It is hard to tell what causing it from the code. Could you please help confirm:
- What are the **old and new** data before/after setDoc with **{merge:true}** ?
- Is there any error message presented when running setDoc? If possible, could you please enable debug log with [firebase.firestore.setLogLevel('debug')](https://firebase.google.com/docs/reference/js/firestore_#setloglevel) and share the log output? You can paste it here (redacting any sensitive information) or email it to me at [milamamat@google.com](mailto:milamamat@google.com).
- Could you please create a minimal reproducible code for this error? Without being able to reproduce it, it is difficult to debug it on our end.

**Comment by raaaahman** [comment]
Thank you for your answer @milaGGL . This is indeed what I wanted to explain with my code sample, although a small mistake was written inside of it: the `merge` option should have been set to `true` but anyway, you got it.

I have tried to set up a minimal reproducible code, but it doesn't seem I am able to reproduce the error... The same goes for my codebase where I found the error, it doesn't seem to happen anymore. I am sorry for the inconvenience.
